megabus/src/main/java/com/bazaarvoice/megabus/refproducer/MegabusRefProducer.java

```diff
@@ -0,0 +1,214 @@
+    private final int _eventsLimit;
+    private final int _skipWaitThreshold;
+
+    private final RateLimitedLog _rateLimitedLog;
+    private final Databus _databus;
+    private final DatabusEventStore _eventStore;
+    private final MetricsGroup _timers;
+    private final String _timerName;
+    private final String _subscriptionName;
+    private final Condition _subscriptionCondition;
+    private final ScheduledExecutorService _executor;
+    private final Producer<String, JsonNode> _producer;
+    private final ObjectMapper _objectMapper;
+    private final Topic _topic;
+    private final Clock _clock;
+
+    public MegabusRefProducer(MegabusRefProducerConfiguration config, Databus databus, DatabusEventStore eventStore, Condition subscriptionCondition,
+                              RateLimitedLogFactory logFactory, MetricRegistry metricRegistry,
+                              Producer<String, JsonNode> producer, ObjectMapper objectMapper, Topic topic,
+                              String partitionIdentifier, String applicationId) {
+        this(config, databus, eventStore, subscriptionCondition, logFactory, metricRegistry, null, producer, objectMapper,
+                topic, partitionIdentifier, applicationId, null);
+    }
+
+    @VisibleForTesting
+    MegabusRefProducer(MegabusRefProducerConfiguration configuration, Databus databus, DatabusEventStore eventStore, Condition subscriptionCondition,
+                       RateLimitedLogFactory logFactory, MetricRegistry metricRegistry,
+                       @Nullable ScheduledExecutorService executor, Producer<String, JsonNode> producer,
+                       ObjectMapper objectMapper, Topic topic, String partitionIdentifer, String applicationId,
+                       Clock clock) {
+
+        _log = LoggerFactory.getLogger(MegabusRefProducer.class.getName() + "-" + partitionIdentifer);
+
+        checkArgument(configuration.getPollIntervalMs() > 0);
+        checkArgument(configuration.getBatchSize() > 0);
+        checkArgument(configuration.getSkipWaitThreshold() >= 0 && configuration.getSkipWaitThreshold() <= configuration.getBatchSize());
+        _pollIntervalMs = configuration.getPollIntervalMs();
+        _eventsLimit = configuration.getBatchSize();
+        _skipWaitThreshold = configuration.getSkipWaitThreshold();
+        _databus = checkNotNull(databus, "databus");
+        _eventStore = checkNotNull(eventStore, "eventStore");
+        _timers = new MetricsGroup(metricRegistry);
+        _timerName = newTimerName("megabusPoll-" + partitionIdentifer);
+        _rateLimitedLog = logFactory.from(_log);
+        _executor = executor;
+        _producer = checkNotNull(producer, "producer");
+        _subscriptionCondition = checkNotNull(subscriptionCondition, "subscriptionCondition");
+        _clock = firstNonNull(clock, Clock.systemUTC());
+
+        //TODO : this is currently hacked to to avoid dedup queues, which require the leader for consistent polling.
```

Comment:
We are not going to do this work in the current branch, as it is going to be _extremely_ complex to modify Emo's dedup leader election system so that I non emo-web server can be leader on a partition. I think it makes sense to leave this TODO so that in case we have throughput issues, but I wouldn't even consider this change until it is our last option to increase throughput.